[frontend] Lazily IR canonicalization and LLVMIR textual generation

erick-xanadu opened the pull request:
**Context:** Canonicalizing the IR before printing is unnecessary but nice to have. However, it increases compile time unnecessarily. Also printing and generating the LLVMIR text form takes time and is not needed.

**Description of the Change:** This PR conditionally lazily canonicalizes the mlir string representation upon request by the user. Also LLVMIR textual representation.

**Benefits:** Less compile time.

**Possible Drawbacks:** `

**Related GitHub Issues:**

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -680,14 +680,16 @@ def generate_ir(self):
-        # TODO: the in-memory and textual form are different after this, consider unification
-        _, mlir_string = canonicalizer.run(mlir_module, self.workspace)
+        mlir_string = None
+        if self.compile_options.keep_intermediate:
```
Hmm, I have certainly used it (as a user) quite a bit outside of testing, as well as in tutorials and demos when referring to or showcasing the IR 🤔 

Does the generating the `.mlir` require the file system?

erick-xanadu replied:
Yes, because we need to write the MLIR representation to a file and then compile it and then read the output file as a string in python.

erick-xanadu replied:
I'm just writing the lazy `.mlir` canonicalizer.

dime10 replied:
> Yes, because we need to write the MLIR representation to a file and then compile it and then read the output file as a string in python.

Limitation of the compiler driver that it can only interface in and out via disk then?

erick-xanadu replied:
I'm fixing that next with shared memory.

erick-xanadu replied:
There is this point of discussion: https://github.com/PennyLaneAI/catalyst/pull/1530/commits/b5a0effbe08a48153da3f0531220e573cdd11328

Now `keep_intermediate=True, target="mlir"` won't generate a file. I think this is acceptable. The new way of achieving the same result would be `target="mlir"` and printing the mlir attribute. But if you think this is not acceptable, then we can cancel this PR or find another alternative.

dime10 replied:
Sounds reasonable to me 👍

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -666,7 +680,7 @@ def closure(qnode, *args, **kwargs):
         return jaxpr, out_type, treedef, dynamic_sig
 
-    @instrument(size_from=0, has_finegrained=True)
+    @instrument(has_finegrained=True)
```
Why did you remove the size instrumentation?

erick-xanadu replied:
It errored out and I thought if there was no `size_from=0` it would just use the only one available. Turns out it expects the return value to be a tuple. So I am fixing that.

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/1530/commits/0383786d970111672d07d2aad8c1030dd963e25f

dime10 replied:
Thanks :)

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -680,16 +694,7 @@ def generate_ir(self):
-        _, mlir_string = canonicalizer.run(mlir_module, self.workspace)
-
-        return mlir_module, mlir_string
+        return mlir_module
```
Worth noting that this changes this signature of this "semi-stable" interface 🤔  One could of course still return the `str(mlir_module)` but maybe we don't want to pay that cost either (and it is somewhat redundant anyways). So it could be worth breaking this now, and I don't think Qrisp directly interfaces with this method?

erick-xanadu replied:
I think it is redundant but it is good to point out in breaking changes. If we need a deprecation cycle, then I still think it is better to remove the string here.

dime10 replied:
Okay yeah they do use it here: https://github.com/eclipse-qrisp/Qrisp/blob/a68c4979009fd5b477903a91dd63651cf8af8f24/src/qrisp/jasp/evaluation_tools/catalyst_interface.py#L139

Might just need to coordinate with them so they can update it.

dime10 replied:
So I think this is fine to change and we can pop them a quick notification that this changed.

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -508,6 +507,21 @@ def __init__(self, fn, compile_options):
+    @property
+    def mlir(self):
+        """obtain the MLIR representation after canonicalization"""
+        # Canonicalize the MLIR since there can be a lot of redundancy coming from JAX.
+        if not self.mlir_module:
+            return None
```
Worth having this cached or no? 

```suggestion        
        if not self.mlir_module:
            return None
        if self.mlir_string:
            return self.mlir_string

        # Canonicalize the MLIR since there can be a lot of redundancy coming from JAX.
```

erick-xanadu replied:
I am not convinced about having it cached. E.g., if there is a recompilation we would show an outdated string.

dime10 replied:
Good point, although I think on recompilation this is easily solved but updating that property as we have always been doing:
https://github.com/PennyLaneAI/catalyst/blob/3340f0c29b6fee19b856189d7c0c2e936c8b9dc4/frontend/catalyst/jit.py#L570-L586
For instance, we could just add this line at the bottom there:
```py
self.mlir_string = None
```

---

Having said that, I think the bug you mention does in fact already exist when you "restore" a previously compiled signature, which doesn't restore the `.mlir` property:
https://github.com/PennyLaneAI/catalyst/blob/3340f0c29b6fee19b856189d7c0c2e936c8b9dc4/frontend/catalyst/jit.py#L592-L597
It also doesn't restore the `.mlir_module` property either, which means even with the proposed design the bug would still exist. Admittedly, it's probably a pretty rare occurrence ...

erick-xanadu replied:
Discussed offline: if you want memoization you will implement it.

erick-xanadu commented:
@dime10 this is just a prototype, but considering we had discussions about printing and how the QIR field would be resolved, I thought prudent to consult you before working further on this. The QIR field is tricky because instrumentation depends on it in order to count the number of lines. Therefore I had to override the default options and add keep_instrumentation to make sure that the LLVM-IR representation is generated when the user requests instrumentation. Again rough draft, will add tests once you approve of the overall design.

This also addresses the comments on the `.mlir` field being empty.

The issue with the "canonical" output (i.e., `catalyst --tool=opt` outputs to stdout vs `catalyst` outputs shared object) is still open and we can open a ticket so that it gets triaged accordingly.

erick-xanadu commented on frontend/catalyst/debug/instruments.py:
```diff
@@ -143,9 +145,14 @@ def wrapper(*args, **kwargs):
+            self = args[0]
+            orig = copy.deepcopy(self.compile_options)
+            self.compile_options.keep_intermediate = True
+            new_args = (self,) + args[1:]
+            fn_results, wall_time, cpu_time = time_function(fn, new_args, kwargs)
```
This line passes the new arguments as opposed to the old arguments. So it will evaluate the fn with the arguments. I.e., your change will revert and invalidate the code above.

dime10 replied:
I don't get what the new arguments are, is `self` not the same as the one already in `args`?

erick-xanadu replied:
I'll give it a try.

erick-xanadu replied:
It worked. [Thanks](https://github.com/PennyLaneAI/catalyst/pull/1530/commits/57aa3db0e07b8ab55ffcefbbb533c07d80bfc171)!

erick-xanadu commented on frontend/catalyst/jit.py:
```diff
@@ -545,14 +559,25 @@ def aot_compile(self):
             )
 
+    @property
+    def qir(self):
```
Whoever is doing this should really use `keep_intermediate=True`. If the IR takes that long, it will not fit in your terminal scrolling and it is better to have a file.

dime10 replied:
Alright, so `keep_intermediate` should populate the `qir` field on first run right? Right now it would run compilation again.

erick-xanadu replied:
> if we get measurable performance improvements.

Also, this depends on the workload. I feel like you are optimizing for the 0.0001% cases. Not doing this change, the QIR benchmark code that I shared on slack that took 12s now takes 120s.

dime10 replied:
> Also, this depends on the workload. I feel like you are optimizing for the 0.0001% cases. Not doing this change, the QIR microbenchmark code that I shared on slack that took 12s now takes 120s.

I'm not optimizing for anything. I'm just saying we should at least be able to measure the difference otherwise the change is not worthwhile. If Lee's example demonstrates that, good. Although to be honest I can't imagine storing massive constant matrices in the IR is really the benchmark we should be using as representative, but it does demonstrate the worst case nicely.

erick-xanadu replied:
> Alright, so keep_intermediate should populate the qir field on first run right? Right now it would run compilation again.

I think we should get rid of the qir field and the mlir fields and just keep keep_intermediate. Since you want to keep these fields, they should at least be lazy since writing to disk takes time and carrying around the IR takes memory. The users using `.qir` and `.mlir` are a small set of users (which is mostly us). This preserves functionality at the cost of increasing time to print `.qir` and `.mlir` which again represents a very small percentage of Catalyst's use.

erick-xanadu replied:
Discussed offline: lazy.

erick-xanadu commented on frontend/catalyst/jit.py:
```diff
@@ -545,14 +559,25 @@ def aot_compile(self):
+            return None
+
+        orig = copy.deepcopy(self.compile_options)
+        self.compile_options.keep_intermediate = True
```
Instead of this, I just added a new variable to the `jit.py` file which does not modify the options. `self.use_cwd_as_workspace` or something like that [here](https://github.com/PennyLaneAI/catalyst/pull/1530/commits/02d81fe4f9b5f50c40db51082f310b383faf5a35)

dime10 replied:
I think you misunderstood my concern:
- the compiler has a configuration to control whether to dump intermediate results (flag `A`)
- the QJIT class has a configuration whether to store intermediate results in the cwd (flag `B`), which since the introduction of flag `A` also implies setting flag `A`

We have two use cases:
- a user setting flag `B`, which means we also internally set `A`
- we internally set `A` (with no effect on `B`)

Right now `A` and `B` share the same name, and I thought it would be cleaner to distinguish between them. Your variable doesn't do that since the user still uses `keep_intermediate` to refer to `B` and we still use `keep_intermediate` to refer to `A`. 

Anyways, since this is a rather minor suggestion on code clarity, it's not a PR blocker.

erick-xanadu commented on doc/releases/changelog-dev.md:
```diff
@@ -49,6 +49,8 @@
     [(#1514)](https://github.com/PennyLaneAI/catalyst/pull/1514)
   - Sort improvements to reduce complexity and memory.
     [(#1524)](https://github.com/PennyLaneAI/catalyst/pull/1524)
+  - Lazily canonicalize the IR.
```
```suggestion
  - Lazy IR canonicalization and LLVMIR textual generation.
```

dime10 reviewed (approved):
Looks good 👍

dime10 commented:
Ah, this reminds me, we should announce this semi-public API breaking changes in the changelog.
